zkevm-circuits/src/evm_circuit/execution/create.rs

```diff
@@ -476,30 +523,44 @@ impl<F: Field> ExecutionGadget<F> for CreateGadget<F> {
             callee_is_persistent.low_u64() != 0,
         )?;
 
+        // retrieve code_hash for creating address
+        let code_hash_previous = block.rws
+            [step.rw_indices[12 + usize::from(is_create2) + copy_rw_increase]]
+            .account_codehash_pair();
+        let code_hash_previous_rlc = region.word_rlc(code_hash_previous.0);
+        self.code_hash_previous
+            .assign(region, offset, code_hash_previous_rlc)?;
+        self.is_code_hash_previous_zero
+            .assign_value(region, offset, code_hash_previous_rlc)?;
+        let is_address_collision = !code_hash_previous.0.is_zero();
+
         let mut rw_offset = 0;
+        if !is_address_collision {
+            let [caller_balance_pair, callee_balance_pair] = if !value.is_zero() {
+                rw_offset += 2;
+                [14, 15].map(|i| {
+                    let rw =
+                        block.rws[step.rw_indices[i + usize::from(is_create2) + copy_rw_increase]];
+                    debug_assert_eq!(
+                        rw.field_tag(),
+                        Some(AccountFieldTag::Balance as u64),
+                        "invalid rw {:?}",
+                        rw
+                    );
+                    rw.account_value_pair()
+                })
+            } else {
+                [(0.into(), 0.into()), (0.into(), 0.into())]
+            };
 
-        let [caller_balance_pair, callee_balance_pair] = if !value.is_zero() {
-            rw_offset += 2;
-            [13, 14].map(|i| {
-                let rw = block.rws[step.rw_indices[i + usize::from(is_create2) + copy_rw_increase]];
-                debug_assert_eq!(
-                    rw.field_tag(),
-                    Some(AccountFieldTag::Balance as u64),
-                    "invalid rw {:?}",
-                    rw
-                );
-                rw.account_value_pair()
-            })
-        } else {
-            [(0.into(), 0.into()), (0.into(), 0.into())]
-        };
-        self.transfer.assign(
-            region,
-            offset,
-            caller_balance_pair,
-            callee_balance_pair,
-            value,
-        )?;
+            self.transfer.assign(
```

Comment:
If I'm not wrong, you will need to assign to the `transfer` gadget some dummy values (zeroes) even when it is a collision, right? So that for those constraints there is no "unassigned cells" ?

Reply:
we have used conditional transfer in CallOp gadget long time ago, it works well.  the reason why don't use dummy value transfer is transfer gadget will add reversible write operation, which cause buss mapping's `handle_reversion` `tx.steps` out of index I remember several months ago tried.